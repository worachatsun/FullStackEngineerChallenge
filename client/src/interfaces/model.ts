export interface IListModel {
    id: number;
    username: string;
    isAdmin: boolean;
    createdAt: Date;
    updatedAt: Date;
}

export interface IReviewModel {
    id: number;
    reviewTo: string;
    reviewBy: string;
    isReview: boolean;
    createdAt: Date;
    updatedAt: Date;
}

export interface IPerformanceModel extends IReviewModel {
    answers: IAnswerModel[];
}

export interface IAnswerModel {
    id: number;
    answer: string;
    reviewId: number;
    questionId: number;
    createdAt: Date;
    UpdatedAt: Date;
}
