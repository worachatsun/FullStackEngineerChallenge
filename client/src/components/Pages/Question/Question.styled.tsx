import styled from 'styled-components';

export const Lable = styled.div`
  font-size: 1.5em;
  font-weight: bold;
`;

export const QuestionContainer = styled.div`
  display: flex;
  justify-content: space-between;
  margin: 15px 0 15px 0;
`;

export const QuestionText = styled.div`
  width: 60%;
`;

export const AnswerContainer = styled.div`
  display: flex;
  justify-content: space-around;
  width: 40%;
`;

export const HeaderContainer = styled(QuestionContainer)`
  height: 40px;
  align-items: center;
  font-size: 1.3em;
  color: #ff0033;
`;

export const ButtonContainer = styled.div`
  display: flex;
  justify-content: flex-end;
  margin-top: 20px;
`;

export const Radio = styled.input``;
