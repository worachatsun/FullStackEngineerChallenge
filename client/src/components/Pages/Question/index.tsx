import React, { FunctionComponent, useEffect } from 'react';
import {
  Lable,
  QuestionContainer,
  QuestionText,
  AnswerContainer,
  ButtonContainer,
  Radio,
  HeaderContainer,
} from './Question.styled';
import { useForm } from 'react-hook-form';
import { IReviewModel } from '../ListReviews';
import Button from '../../commons/Button';
import { fetcher, mutator, HttpMethod } from '../../commons/utils/client';
import {
  LIST_QUESTIONS_API,
  ANSWER_API,
  REVIEW_LIST_API,
} from '../../../constants/routes';
import useSWR, { mutate } from 'swr';

interface IProps {
  review: IReviewModel;
  close: () => void;
}

interface IQuestionList {
  question: string;
  id: number;
}

const Question: FunctionComponent<IProps> = ({ review, close }) => {
  const { register, handleSubmit, errors } = useForm();
  const token = localStorage.getItem('token');
  const { data: questions } = useSWR<IQuestionList[]>(
    [LIST_QUESTIONS_API, token],
    fetcher
  );

  const onSubmit = async (reviews: any) => {
    let answers: any[] = [];
    Object.entries(reviews).forEach(([questionId, answer]) => {
      answers.push({ questionId, answer, reviewId: review.id });
    });
    try {
      const { response } = await mutator(
        ANSWER_API,
        HttpMethod.POST,
        token as string,
        {
          answers,
        }
      );
      if (!response?.ok)
        throw new Error(`${response?.status}: ${response?.statusText}`);
      close();
    } catch (error) {
      console.error(error);
    }
  };

  return questions ? (
    <>
      <Lable>Review to {review.reviewTo}</Lable>
      <HeaderContainer>
        <QuestionText>Questions</QuestionText>
        <AnswerContainer>
          <div>1</div>
          <div>2</div>
          <div>3</div>
          <div>4</div>
          <div>5</div>
        </AnswerContainer>
      </HeaderContainer>
      <form onSubmit={handleSubmit(onSubmit)}>
        {questions.map(({ question, id }, index: number) => (
          <QuestionContainer key={index}>
            <QuestionText>{question}</QuestionText>
            <AnswerContainer>
              <Radio
                type='radio'
                value='1'
                name={`${id}`}
                ref={register({ required: true })}
              />
              <Radio
                type='radio'
                value='2'
                name={`${id}`}
                ref={register({ required: true })}
              />
              <Radio
                type='radio'
                value='3'
                name={`${id}`}
                ref={register({ required: true })}
              />
              <Radio
                type='radio'
                value='4'
                name={`${id}`}
                ref={register({ required: true })}
              />
              <Radio
                type='radio'
                value='5'
                name={`${id}`}
                ref={register({ required: true })}
              />
            </AnswerContainer>
          </QuestionContainer>
        ))}
        <ButtonContainer>
          <Button>Submit</Button>
        </ButtonContainer>
      </form>
    </>
  ) : (
    <div></div>
  );
};

export default Question;
