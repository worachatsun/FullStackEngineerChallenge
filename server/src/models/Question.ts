import { Model, DataTypes } from 'sequelize';
import { sequelize } from '../loaders/mysql';

export class Question extends Model {
  id!: number;
  question!: string;
}

Question.init(
  {
    question: {
      type: DataTypes.STRING,
      unique: true,
      allowNull: false,
    },
    createdAt: {
      type: DataTypes.DATE,
      defaultValue: new Date(),
      allowNull: false,
    },
    updatedAt: {
      type: DataTypes.DATE,
      defaultValue: new Date(),
      allowNull: false,
    },
  },
  {
    sequelize,
    tableName: 'Question',
    freezeTableName: true,
  }
);
