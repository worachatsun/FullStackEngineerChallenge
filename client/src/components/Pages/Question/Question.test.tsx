import React from "react";
import renderer from "react-test-renderer";
import Question from "./index";
import { IPerformanceModel } from "../../../interfaces/model";

describe("<Question />", () => {
    const review = {
        id: 1,
        reviewTo: "sun",
        reviewBy: "sun",
        isReview: false,
        createdAt: new Date(),
        updatedAt: new Date(),
        answers: [],
    };
    test("should render correctly", () => {
        const render = renderer.create(<Question review={review as IPerformanceModel} close={jest.fn} />).toJSON();
        expect(render).toMatchSnapshot();
    });
});
