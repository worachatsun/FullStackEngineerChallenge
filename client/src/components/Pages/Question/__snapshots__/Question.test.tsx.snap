// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Question /> should render correctly 1`] = `<div />`;
